image/vision_transformer/colo_vit/test_vit.py

```diff
@@ -0,0 +1,114 @@
+import torch
+import pytest
+import colossalai
+from colossalai.context.parallel_mode import ParallelMode
+import torch.multiprocessing as mp
+from colossalai.testing import rerun_if_address_is_in_use
+from colossalai.utils.cuda import get_current_device
+from colossalai.utils import free_port
+from colossalai.utils.model.colo_init_context import ColoInitContext
+from colossalai.tensor import TensorSpec, ComputePattern, ParallelAction, DistSpecManager, distspec
+from colossalai.core import global_context as gpc
+from functools import partial
+from utils.util import tensor_equal, tensor_shard_equal, set_seed
+from vit import get_training_components
+from torch.nn.parallel import DistributedDataParallel as DDP
+from colossalai.nn.parallel.data_parallel import ColoDDP
+
+
+def init_1d_row_spec(model):
+    spec = TensorSpec(
+        distspec.shard(gpc.get_group(ParallelMode.PARALLEL_1D), [-1], [gpc.get_world_size(ParallelMode.PARALLEL_1D)]),
+        ParallelAction(ComputePattern.TP1D))
+    with DistSpecManager.no_grad():
+        for n, p in model.named_parameters():
+            if 'weight' in n and 'layernorm' not in n and 'embeddings.patch_embeddings.projection.weight' not in n:
```

Comment:
The function name is not clear though it will not make a wrong result since every output is gathered if the gather_out arg is not set. In fact, this function sets TP_1D for dim=-1(the last dim). Only for all Linear, it's 1d_row split because Linear will be transposed when calculating. But for other layers, it's 1d_col split.

Reply:
The result is incorrect with 'projection.weight'.
I follow the implementation from tests/test_tensor/test_gpt.py to get test_vit.py
I am not familiar with detailed implementation of ColoTensor, but it seems that it can pass all the tests.
Any suggestions for further revision?

Reply:
e.g. to find out why projection.weight doesn't work?